unit/test-rilmodem-cf.c

```diff
@@ -0,0 +1,513 @@
+{
+	struct rilmodem_cf_data *rcd = data;
+	struct ofono_call_forwarding *cf = g_malloc0(sizeof(*cf));
+	int retval;
+
+	retval = cfdriver->probe(cf, OFONO_RIL_VENDOR_AOSP, rcd->ril);
+	g_assert(retval == 0);
+
+	return cf;
+}
+
+void ofono_call_forwarding_register(struct ofono_call_forwarding *cf)
+{
+}
+
+void ofono_call_forwarding_remove(struct ofono_call_forwarding *cf)
+{
+}
+
+void ofono_call_forwarding_set_data(struct ofono_call_forwarding *cf,
+					void *data)
+{
+	cf->driver_data = data;
+}
+
+void *ofono_call_forwarding_get_data(struct ofono_call_forwarding *cf)
+{
+	return cf->driver_data;
+}
+
+/*
+ * As all our architectures are little-endian except for
+ * PowerPC, and the Binder wire-format differs slightly
+ * depending on endian-ness, the following guards against test
+ * failures when run on PowerPC.
+ */
+#if BYTE_ORDER == LITTLE_ENDIAN
+
+static GMainLoop *mainloop;
+
+static void cf_set_callback(const struct ofono_error *error, void *data)
+{
+	struct rilmodem_cf_data *rcd = data;
+	const struct cf_data *cfd = rcd->test_data;
+
+	g_assert(error->type == cfd->error_type);
+
+	g_main_loop_quit(mainloop);
+}
+
+static void cf_query_callback(const struct ofono_error *error, int total,
+			const struct ofono_call_forwarding_condition *list,
+			void *data)
+{
+	struct rilmodem_cf_data *rcd = data;
+	const struct cf_data *cfd = rcd->test_data;
+
+	g_assert(error->type == cfd->error_type);
+
+	if (error->type == OFONO_ERROR_TYPE_NO_ERROR) {
+		int i;
+
+		g_assert(total == cfd->total);
+
+		for (i = 0; i < total; ++i) {
+			const struct ofono_call_forwarding_condition *cur, *cmp;
+
+			cur = &list[i];
+			cmp = &cfd->list[i];
+			g_assert(cur->status == cmp->status);
+			g_assert(cur->cls == cmp->cls);
+			g_assert_cmpstr(cur->phone_number.number, ==,
+						cmp->phone_number.number);
+			g_assert(cur->phone_number.type
+						== cmp->phone_number.type);
+			g_assert(cur->time == cmp->time);
+		}
+	}
+
+	g_main_loop_quit(mainloop);
+}
+
+static gboolean trigger_cf_call(struct rilmodem_cf_data *rcd)
+{
+	switch(rcd->test_data->call_type) {
+	case CF_CALL_ACTIVATION:
+		cfdriver->activation(rcd->cf, rcd->test_data->type,
+						rcd->test_data->cls,
+						cf_set_callback, rcd);
+		break;
+	case CF_CALL_REGISTRATION:
+		cfdriver->registration(rcd->cf, rcd->test_data->type,
+						rcd->test_data->cls,
+						&rcd->test_data->number,
+						rcd->test_data->time,
+						cf_set_callback, rcd);
+		break;
+	case CF_CALL_DEACTIVATION:
+		cfdriver->deactivation(rcd->cf, rcd->test_data->type,
+						rcd->test_data->cls,
+						cf_set_callback, rcd);
+		break;
+	case CF_CALL_ERASURE:
+		cfdriver->erasure(rcd->cf, rcd->test_data->type,
+						rcd->test_data->cls,
+						cf_set_callback, rcd);
+		break;
+	case CF_CALL_QUERY:
+		cfdriver->query(rcd->cf, rcd->test_data->type,
+				rcd->test_data->cls, cf_query_callback, rcd);
+		break;
+	};
+
+	return FALSE;
+}
+
+/* Test SET_CALL_FORWARD. Real transaction from turbo. */
```

Comment:
Please add a short description of what the real transaction was ( ie. register unconditional, not-reachable, ... ).  This type of comment is included in other tests.